src/ROSInterface.cpp

```diff
@@ -0,0 +1,348 @@
+/*
+ * Copyright (c) 2022, Daniel Beßler
+ * All rights reserved.
+ *
+ * This file is part of KnowRob, please consult
+ * https://github.com/knowrob/knowrob for license details.
+ */
+
+#include <termios.h>
+// STD
+#include <exception>
+#include <iostream>
+// BOOST
+#include <boost/program_options/options_description.hpp>
+#include <boost/program_options/variables_map.hpp>
+#include <boost/program_options/parsers.hpp>
+#include <boost/property_tree/json_parser.hpp>
+#include <boost/property_tree/ptree.hpp>
+// KnowRob
+#include <knowrob/knowrob.h>
+#include <knowrob/logging.h>
+#include <knowrob/HybridQA.h>
+#include <knowrob/queries.h>
+// ROS
+#include <ros/ros.h>
+#include <ros/console.h>
+#include <ros/package.h>
+#include <json_prolog_msgs/PrologQuery.h>
+#include <json_prolog_msgs/PrologFinish.h>
+#include <json_prolog_msgs/PrologNextSolution.h>
+
+#define PARAM_INITIAL_PACKAGE "initial_package"
+#define PARAM_INITIAL_GOAL "initial_goal"
+#define PARAM_NUM_PL_THREADS "num_pl_threads"
+#define PARAM_NUM_ROS_THREADS "num_ros_threads"
+#define PARAM_ENABLE_DEBUG "pl_debug"
+
+#define NUM_PL_THREADS_DEFAULT 2
+#define NUM_ROS_THREADS_DEFAULT 2
+
+using namespace knowrob;
+namespace po = boost::program_options;
+
+static const char* PROMPT = "?- ";
+
+// TODO support queries like: mongolog:(woman(X), ...) or (mongolog | prolog):(woman(X), ...)
+//    but there could be different instances of the same reasoner type. So might need a reasoner id instead.
+
+class HybridQAResultHandler : public QueryResultHandler {
+public:
+    bool incrementalMode;
+    std::string errorMessage;
+    std::list<std::shared_ptr<std::string>> solutions_;
+    HybridQAResultHandler(const boost::property_tree::ptree &config, const std::string &queryString, bool incremental = false)
+	: has_stop_request_(false),
+	  currentQuery_(""),
+      hybridQA_(config)
+	{
+        incrementalMode = incremental;
+        errorMessage = "";
+        currentQuery_ = queryString;
+        runQuery(currentQuery_);
+	}
+
+    std::string substitutionToString(const QueryResultPtr &sharedPtr) {
+        return "";
+    }
+
+// Override QueryResultHandler
+    bool pushQueryResult(const QueryResultPtr &solution) override {
+        if(solution->substitution()->empty()) {
+            solutions_.push_back(std::make_shared<std::string>("True."));
+        }
+        else {
+            std::stringstream ss;
+            ss << *solution->substitution() << ".";
+            solutions_.push_back(std::make_shared<std::string>(ss.str()));
+        }
+        numSolutions_ += 1;
+        return !has_stop_request_;
+    }
+
+    bool has_more_solutions() {
+        return numSolutions_ > 0;
+    }
+
+    std::string next_solution() {
+        std::string solution = solutions_.front()->c_str();
```

Comment:
remove *c_str()* it returns a `char*` and through assignment you convert it to *std::string* again, so the call is effectively not needed